dissect/cstruct/types/structure.py

```diff
@@ -1,69 +1,146 @@
-    """Holds a structure field."""
+    """Structure field."""
 
-    def __init__(self, name: str, type_: BaseType, bits: int = None, offset: int = None):
+    def __init__(self, name: str, type_: MetaType, bits: int = None, offset: int = None):
         self.name = name
         self.type = type_
         self.bits = bits
         self.offset = offset
-        self.alignment = type_.alignment
+        self.alignment = type_.alignment or 1
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         bits_str = f" : {self.bits}" if self.bits else ""
-        return f"<Field {self.name} {self.type}{bits_str}>"
-
-
-class Structure(BaseType):
-    """Type class for structures."""
-
-    def __init__(
-        self, cstruct: cstruct, name: str, fields: List[Field] = None, align: bool = False, anonymous: bool = False
-    ):
-        super().__init__(cstruct)
-        self.name = name
-        self.size = None
-        self.alignment = None
-
-        self.lookup = OrderedDict()
-        self.fields = fields
-
-        self.align = align
-        self.anonymous = anonymous
-        self.dynamic = False
+        return f"<Field {self.name} {self.type.__name__}{bits_str}>"
+
+
+class StructureMetaType(MetaType):
+    """Base metaclass for cstruct structure type classes."""
+
+    # TODO: resolve field types in _update_fields, remove resolves elsewhere?
+
+    fields: dict[str, Field]
+    """Mapping of field names to :class:`Field` objects, including "folded" fields from anonymous structures."""
+    lookup: dict[str, Field]
+    """Mapping of "raw" field names to :class:`Field` objects. E.g. holds the anonymous struct and not its fields."""
+    __fields__: list[Field]
+    """List of :class:`Field` objects for this structure. This is the structures' Single Source Of Truth."""
+
+    # Internal
+    __align__: bool
+    __anonymous__: bool
+    __updating__ = False
+    __compiled__ = False
+
+    def __new__(metacls, cls, bases, classdict: dict[str, Any]) -> MetaType:
+        if (fields := classdict.pop("fields", None)) is not None:
+            metacls._update_fields(metacls, fields, align=classdict.get("align", False), classdict=classdict)
+
+        return super().__new__(metacls, cls, bases, classdict)
+
+    def __call__(cls, *args, **kwargs) -> Structure:
+        if (
+            cls.__fields__
+            and len(args) == len(cls.__fields__) == 1
+            and isinstance(args[0], bytes)
+            and issubclass(cls.__fields__[0].type, bytes)
+            and len(args[0]) == cls.__fields__[0].type.size
+        ):
+            # Shortcut for single char/bytes type
+            return type.__call__(cls, *args, **kwargs)
+        elif not args and not kwargs:
+            obj = cls(**{field.name: field.type.default() for field in cls.__fields__})
+            object.__setattr__(obj, "_values", {})
+            object.__setattr__(obj, "_sizes", {})
+            return obj
+
+        return super().__call__(*args, **kwargs)
```

Comment:
```suggestion
    def __call__(cls, *args, **kwargs) -> Structure:
        result = None
        if (
            cls.__fields__
            and len(args) == len(cls.__fields__) == 1
            and isinstance(args[0], bytes)
            and issubclass(cls.__fields__[0].type, bytes)
            and len(args[0]) == cls.__fields__[0].type.size
        ):
            # Shortcut for single char/bytes type
            result = type.__call__(cls, *args, **kwargs)
        elif not args and not kwargs:
            obj = cls(**{field.name: field.type.default() for field in cls.__fields__})
            object.__setattr__(obj, "_values", {})
            object.__setattr__(obj, "_sizes", {})
            result = obj
        else:
            result = super().__call__(*args, **kwargs)
        return result

```

Might be handy here to have a single return point.

Reply:
I prefer this style in this specific case.